vm/src/exceptions.rs

```diff
@@ -538,24 +540,62 @@ fn import_error_init(vm: &VirtualMachine, args: PyFuncArgs) -> PyResult {
     Ok(vm.get_none())
 }
 
-pub fn init(context: &PyContext) {
-    let base_exception_type = &context.exceptions.base_exception_type;
-    PyBaseException::extend_class(context, base_exception_type);
+fn none_getter(_obj: PyObjectRef, vm: &VirtualMachine) -> PyObjectRef {
+    vm.get_none()
+}
+
+fn make_arg_getter(idx: usize) -> impl Fn(PyBaseExceptionRef, &VirtualMachine) -> PyResult {
+    move |exc, vm| {
+        exc.args
+            .borrow()
+            .elements
+            .get(idx)
+            .cloned()
+            .ok_or_else(|| vm.new_value_error(format!("couldn't get arg {} of exception", idx)))
+    }
+}
+
+pub fn init(ctx: &PyContext) {
+    let excs = &ctx.exceptions;
 
-    let exception_type = &context.exceptions.exception_type;
-    extend_class!(context, exception_type, {
-        "__str__" => context.new_rustfunc(exception_str),
-        "__repr__" => context.new_rustfunc(exception_repr),
+    PyBaseException::extend_class(ctx, &excs.base_exception_type);
+
+    extend_class!(ctx, &excs.exception_type, {
+        "__str__" => ctx.new_rustfunc(exception_str),
+        "__repr__" => ctx.new_rustfunc(exception_repr),
     });
 
-    let import_error_type = &context.exceptions.import_error;
-    extend_class!(context, import_error_type, {
-        "__init__" => context.new_rustfunc(import_error_init)
+    extend_class!(ctx, &excs.import_error, {
+        "__init__" => ctx.new_rustfunc(import_error_init)
     });
 
-    extend_class!(context, &context.exceptions.stop_iteration, {
-        "value" => context.new_rustfunc(|obj: PyObjectRef, vm: &VirtualMachine| {
+    extend_class!(ctx, &excs.stop_iteration, {
+        "value" => ctx.new_rustfunc(|obj: PyObjectRef, vm: &VirtualMachine| {
             objiter::stop_iter_value(vm, &obj)
         }),
     });
+
+    extend_class!(ctx, &excs.unicode_decode_error, {
```

Comment:
Can we add some tests for this magic? Why is this done in Rust and not in Python?

Reply:
What do you mean? It's properties on a built-in type, there's no reason to do it in Python. It's not really magic, just deduplicating getter code that would have been repetitive otherwise.